src/cli/completion.rs

```diff
@@ -0,0 +1,160 @@
+use crate::cli::{Args, CompletionCommand};
+use clap::CommandFactory;
+use miette::IntoDiagnostic;
+use regex::Regex;
+use std::borrow::Cow;
+use std::io::Write;
+use std::str::from_utf8_mut;
+
+/// Generate completions for the pixi cli, and print those to the stdout
+pub(crate) fn execute(args: CompletionCommand) -> miette::Result<()> {
+    let clap_shell = args
+        .shell
+        .or(clap_complete::Shell::from_env())
+        .unwrap_or(clap_complete::Shell::Bash);
+
+    let mut script = vec![];
+
+    // Generate the original completion script.
+    clap_complete::generate(clap_shell, &mut Args::command(), "pixi", &mut script);
+
+    let script = match clap_shell {
+        clap_complete::Shell::Bash => {
+            replace_bash_completion(from_utf8_mut(&mut script).into_diagnostic()?)
+        }
+        clap_complete::Shell::Zsh => {
+            replace_zsh_completion(from_utf8_mut(&mut script).into_diagnostic()?)
+        }
+        _ => Cow::Borrowed(from_utf8_mut(&mut script).into_diagnostic()?),
+    };
+
+    std::io::stdout()
+        .write_all(script.as_bytes())
+        .into_diagnostic()?;
+
+    Ok(())
+}
+
+fn replace_bash_completion(script: &str) -> Cow<str> {
+    let pattern = r#"(?s)pixi__run\).*?opts="(.*?)".*?(if.*?fi)"#;
+    // NOTE THIS IS FORMATTED BY HAND
+    let replacement = r#"pixi__run)
+            opts="$1"
+            if [[ $${cur} == -* ]] ; then
+               COMPREPLY=( $$(compgen -W "$${opts}" -- "$${cur}") )
+               return 0
+            elif [[ $${COMP_CWORD} -eq 2 ]]; then
+               local tasks=$$(pixi task list --summary 2> /dev/null)
+               if [[ $$? -eq 0 ]]; then
+                   COMPREPLY=( $$(compgen -W "$${tasks}" -- "$${cur}") )
+                   return 0
+               fi
+            fi"#;
+    let re = Regex::new(pattern).unwrap();
+    re.replace(script, replacement)
+}
+
+fn replace_zsh_completion(script: &str) -> Cow<str> {
+    let pattern = r#"(?ms)(\(run\))(?:.*?)(_arguments.*?)(\*::task)"#;
+
+    // NOTE THIS IS FORMATTED BY HAND
+    let zsh_replacement = r#"$1
+_values 'task' $$( pixi task list --summary 2> /dev/null )
+$2::task"#;
+
+    let re = Regex::new(pattern).unwrap();
+    re.replace(script, zsh_replacement)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    pub fn test_zsh_completion() {
```

Comment:
Instead of hardcoding this script you could also take the output from `clap_completion`! By first generating the clap completion script.

Reply:
Yes good point, added a test that acts on the clap complete. We can't really test if the exact replacement is done correctly as the file can always update by other additions. So I'm now at least testing if the regex replaces anything which should only happen when there was a match.